### PR TITLE
[InstCombine] Avoid crash on aggregate types in SimplifyDemandedUseFPClass

### PR DESCRIPTION
**Pull request by MacDue**
The disables folding for FP aggregates that are not poison/posZero types, which is currently not supported. Note: To fully handle this aggregates would also likely require teaching `computeKnownFPClass()` to handle array and struct constants (which does not seem implemented outside of zero init).

### CODE REVIEW

**Comment by MacDue** [comment]
@mikaelholmen could you test if this resolves the issue you reported? (https://github.com/llvm/llvm-project/pull/110506#issuecomment-2392839509) :slightly_smiling_face:

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 1]
```diff
@@ -1919,6 +1919,19 @@ Value *InstCombinerImpl::SimplifyDemandedVectorElts(Value *V,
 /// For floating-point classes that resolve to a single bit pattern, return that
 /// value.
 static Constant *getFPClassConstant(Type *Ty, FPClassTest Mask) {
+  // TODO: Support aggregate types that are allowed by FPMathOperator.
+  switch (Mask) {
+    case fcPosZero:
+    case fcNegZero:
+    case fcPosInf:
+    case fcNegInf:
+      if (Ty->isAggregateType())
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:1928]: fcPosZero can just use ConstantAggregateZero

File: llvm/test/Transforms/InstCombine/simplify-demanded-fpclass.ll [hunk 2]
```diff
@@ -91,6 +91,35 @@ define nofpclass(inf) float @ret_nofpclass_inf__ninf() {
...
 }
 
+; Basic aggregate tests to ensure this does not crash.
+define nofpclass(nan) { float } @ret_nofpclass_struct_ty() {
+; CHECK-LABEL: define nofpclass(nan) { float } @ret_nofpclass_struct_ty() {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    ret { float } zeroinitializer
+;
+entry:
+  ret { float } zeroinitializer
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/simplify-demanded-fpclass.ll:101]: Test double element and struct of vector

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 3]
```diff
@@ -1919,6 +1919,19 @@ Value *InstCombinerImpl::SimplifyDemandedVectorElts(Value *V,
 /// For floating-point classes that resolve to a single bit pattern, return that
 /// value.
 static Constant *getFPClassConstant(Type *Ty, FPClassTest Mask) {
+  // TODO: Support aggregate types that are allowed by FPMathOperator.
+  switch (Mask) {
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:1923]: It wasn't obvious to me why these pre-cases. Merge the switches?

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 4]
```diff
@@ -1919,6 +1919,19 @@ Value *InstCombinerImpl::SimplifyDemandedVectorElts(Value *V,
...
+    case fcPosZero:
+    case fcNegZero:
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:1925]: These cases aren't tested

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 5]
```diff
@@ -1919,6 +1919,19 @@ Value *InstCombinerImpl::SimplifyDemandedVectorElts(Value *V,
...
+    case fcPosInf:
+    case fcNegInf:
+      if (Ty->isAggregateType())
+        return nullptr;
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:1929]: It's okay to do this unconditionally, without repeating the switch.

**Comment by mikaelholmen** [comment]
> @mikaelholmen could you test if this resolves the issue you reported? ([#110506 (comment)](https://github.com/llvm/llvm-project/pull/110506#issuecomment-2392839509)) 🙂

Indeed it does. Thank you!

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp (hunk 1 above)
Comment (MacDue) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp]: Done :+1:

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp (hunk 3 above)
Comment (MacDue) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp]: Refactored now, but I just didn't want to repeat the `isAggregateType()` check for every case.

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp (hunk 4 above)
Comment (MacDue) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp]: I've put cases that (I think) _should_ hit these, however, `computeKnownFPClass()` only handles `ConstantAggregateZero` (and that case is now fixed), so I'm not sure it's possible to get fcPosZero, fcNegZero, etc for a constant aggregate.

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp [hunk 6]
```diff
@@ -1919,17 +1919,26 @@ Value *InstCombinerImpl::SimplifyDemandedVectorElts(Value *V,
...
-  switch (Mask) {
-  case fcPosZero:
+  if (Mask == fcNone)
+    return PoisonValue::get(Ty);
+
+  if (Mask == fcPosZero) {
+    if (Ty->isAggregateType())
+      return ConstantAggregateZero::get(Ty);
     return ConstantFP::getZero(Ty);
+  }
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:1929]:
```suggestion
  if (Mask == fcPosZero) {
    return Constant::getNullValue(Ty);
```

Comment (MacDue) [reply llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp]: Nice, thanks :+1:

**Review by nikic — APPROVED** [review]

**Pull request merged**